### PR TITLE
DescribeParser: Support concurrent befores, afters, and children

### LINKED ISSUE #93: concurrency issue: Internal Error occured.
It doesn't do it every time but running the following spek sometimes causes the below NPE.

``` kotlin
class ParallelSpek : Spek({
    given("some parallel operations") {
        // uses `parallelStream` from `kotlinx-support-jdk8-0.1`
        listOf(1, 2, 3, 4, 5).parallelStream().forEach { int ->
            on("execution") {
                it("should something") {}
                it("should something else") { fail() }
                it("should yet something else") { check(false) }
            }
        }
    }
})
```

```
Internal Error occured.
java.lang.NullPointerException
    at java.util.LinkedList$ListItr.next(LinkedList.java:893)
    at org.jetbrains.spek.junit.JUnitDescriptionCache.create(JUnitDescriptionCache.kt:45)
    at org.jetbrains.spek.junit.JUnitDescriptionCache.get(JUnitDescriptionCache.kt:20)
    at org.jetbrains.spek.junit.JUnitDescriptionCache.create(JUnitDescriptionCache.kt:30)
    at org.jetbrains.spek.junit.JUnitDescriptionCache.get(JUnitDescriptionCache.kt:20)
    at org.jetbrains.spek.junit.JUnitSpekRunner.getDescription(JUnitSpekRunner.kt:23)
    at com.intellij.junit4.JUnit4IdeaTestRunner.getDescription(JUnit4IdeaTestRunner.java:123)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:99)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
    at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
    at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
```

### PR DESCRIPTION
**Pull request by mfulton26**
Fixes #93.

### CODE REVIEW

File: spek-core/src/main/kotlin/org/jetbrains/spek/api/DescribeParser.kt [hunk 1]
```diff
@@ -1,11 +1,11 @@
 package org.jetbrains.spek.api
 
-import java.util.*
+import java.util.concurrent.CopyOnWriteArrayList
 
 class DescribeParser() : DescribeBody {
-    val befores: LinkedList<() -> Unit> = LinkedList()
```
Comment (artem-zinnatullin) [inline spek-core/src/main/kotlin/org/jetbrains/spek/api/DescribeParser.kt]: Why it was linked…?

Comment (mfulton26) [inline spek-core/src/main/kotlin/org/jetbrains/spek/api/DescribeParser.kt]: Good question. I couldn't see any apparent reason for it.

**Comment by raniejade** [comment]
Spek is going under some significant changes before `1.0`, this may or may not be applicable with the new changes.

**Comment by mfulton26** [comment]
Oh, okay, good to know. Thanks.

FYI: I accidentally committed failing tests so I amended my changed and pushed it.

Thanks.

**Comment by hhariri** [comment]
Given the new structure, I believe for now this won't be applicable. The idea though can be implemented moving forward after 1.0 release. Thanks!

**Pull request closed without merging**